composer/algorithms/utils/augmentation_common.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from typing import Type, TypeVar, cast
+
+import torch
+from PIL.Image import Image as PillowImage
+from torchvision import transforms
+
+_InputImgT = TypeVar("_InputImgT", torch.Tensor, PillowImage)
+_OutputImgT = TypeVar("_OutputImgT", torch.Tensor, PillowImage)
+
+
+def image_as_type(image: _InputImgT, need_type: Type[_OutputImgT]) -> _OutputImgT:
+    """Creates a copy of an image-like object with a different type.
+
+    Args:
+        image: a tensor or PIL image. Batches of images in NCHW format are
+            also supported.
+        need_type: type of the copied image
+
+    Returns:
+        A copy of ``image`` with type ``need_type``.
+
+    Raises:
+        RuntimeError: if ``image`` is batch of images, but ``need_type`` is
+            a PIL image. We could convert to a batch of PIL images, but we
+            instead fail fast since this is not expected behavior at any
+            call sites (within data augmenation functions).
+        TypeError: if ``need_type`` is not one of :class:`torch.Tensor` or
+            :class:`PIL.Image.Image`.
+    """
+    if isinstance(image, need_type):
+        return image
+
+    # fail fast if attempting to convert a batch of images to a single PIL image
+    is_batch = isinstance(image, torch.Tensor) and (image.ndim == 4) and (image.shape[0] > 1)
+    if is_batch and issubclass(need_type, PillowImage):
+        raise RuntimeError(f"Could not convert batch of images to PIL image")
+
+    if not issubclass(need_type, (torch.Tensor, PillowImage)):
+        raise TypeError(f"Only need_type={{torch.Tensor, Image}} is supported; got {need_type}")
+
+    if isinstance(image, PillowImage):
+        return transforms.functional.to_tensor(image)  # PIL -> Tensor
+    return transforms.functional.to_pil_image(image)  # Tensor -> PIL
+
+
+def image_typed_and_shaped_like(image: _InputImgT,
```

Comment:
Do we need this? can't the user just do:
```
new_image = image_as_type(old_image, type=type(reference_image))
```

? What is the need for all the typecasting behavior?

Reply:
The typecasting is purely to appease pyright; this was a much shorter function without it. Calling `image_as_type` is not sufficient because the shape doesn't get changed.

Reply:
OK I see, that's unfortunate. Is there a reason we only reshape when `reference_image` is of type `Tensor`?

Reply:
managed to rm this whole function